soccer/optimization/ParallelGradientAscent1D.cpp

```diff
@@ -0,0 +1,74 @@
+#include "ParallelGradientAscent1D.hpp"
+#include <math.h>
+#include <iostream>
+
+ParallelGradientAscent1D::ParallelGradientAscent1D(
+    ParallelGradient1DConfig* config)
+    : config(config) {
+    // Create list of GA1Ds
+    for (int i = 0; i < (config->GA1DConfig).size(); i++) {
+        GA1Ds.push_back(GradientAscent1D(&config->GA1DConfig.at(i)));
+    }
+}
+
+/**
+ * Executes all GA1Ds until their max has been reached
+ */
+void ParallelGradientAscent1D::execute() {
+    // While any are not done
+    bool continueExecution = true;
+
+    while (continueExecution) {
+        // Default to false unless any still need to work
+        continueExecution = false;
+
+        // Execute a step for each one
+        for (auto& GA1D : GA1Ds) {
+            if (GA1D.continueExecution()) {
+                GA1D.singleStep();
+                continueExecution = true;
+            }
+        }
+
+        // Assume ascending order for xStart
+        for (int i = 0; i < GA1Ds.size() - 1; i++) {
+            GradientAscent1D lower = GA1Ds.at(i);
+            GradientAscent1D upper = GA1Ds.at(i + 1);
+
+            // Erase elements if they get too close
+            // This helps kill any GA1Ds that are going up the same hill
+            if (fabs(lower.getXValue() - upper.getXValue()) <
+                config->xCombineThresh) {
+                GA1Ds.erase(GA1Ds.begin() + i + 1);
```

Comment:
Oh woops. I didn't realize what was happening exactly. Um it would be a bit more complicated and sort of dumb, but you could do like.
auto i=1;
GA1Ds.erase(remove_if(GA1Ds.begin(), GA1Ds.end(), [&GA1Ds] (const auto &value) {
auto&next = GA1Ds[i];
i++;
return fabs(value.getXValue() - next .getXValue()) < config->xCombineThresh);
}
Might technically be faster since not everything gets copied each iteration. Not actually sure if it matters though since idk what the size of GAID is.